ledger/acctupdates.go

```diff
@@ -878,14 +897,80 @@ func (au *accountUpdates) lookupWithRewards(rnd basics.Round, addr basics.Addres
 	var rewardsProto config.ConsensusParams
 	var rewardsLevel uint64
 	var persistedData persistedAccountData
+	var persistedResources []persistedResourcesData
+	var resourceDbRound basics.Round
 	withRewards := true
+
+	foundAccount := false
+	var ad ledgercore.AccountData
+
+	var foundResources map[basics.CreatableIndex]basics.Round
+	var resourceCount uint64
+
+	addResource := func(cidx basics.CreatableIndex, round basics.Round, res ledgercore.AccountResource) {
+		foundRound, ok := foundResources[cidx]
+		if !ok { // first time seeing this cidx
+			resourceCount++
+			foundResources[cidx] = round
+			res.AssignAccountData(&data)
+			return
+		}
+		// is this newer than current "found" rnd for this resource? (XXX is this not possible?)
+		if round > foundRound {
+			panic("logic error in lookupAllResources") // remove after testing
+		}
+		// otherwise older than current "found" rnd: ignore, since it's older than what we have
+	}
+
+	// possibly avoid a trip to the DB for more resources, if we can use totals info
+	checkDone := func() bool {
+		if foundAccount { // found AccountData
+			// no resources
+			if ad.TotalAssetParams == 0 && ad.TotalAppParams == 0 &&
+				ad.TotalAssets == 0 && ad.TotalAppLocalStates == 0 {
+				return true
+			}
+			// not possible to know how many resources rows to look for: totals conceal possibly overlapping assets/apps
+			if (ad.TotalAssetParams != 0 && ad.TotalAssets != 0) ||
+				(ad.TotalAppParams != 0 && ad.TotalAppLocalStates != 0) {
+				return false
+			}
+
+			// in cases where acct is only a holder of assets/apps, or is just a creator of assets/apps,
+			// we can know how many resources rows to look for
+			needToFind := uint64(0)
+			if ad.TotalAssetParams == 0 { // not a creator of assets
+				needToFind += uint64(ad.TotalAssets) // look for N asset holdings
+			} else if ad.TotalAssets == 0 { // not a holder of assets
+				needToFind += uint64(ad.TotalAssetParams) // look for N asset params
```

Comment:
Being asset creator assumes to have asset holding as well. That's why in many places only `ad.Assets` are checked (so a combination of ad.TotalAssetParams != 0 and ad.TotalAssets == 0 is not possible.

If user A created 5 assets and opted in into 2 other, so it has
TotalAssetParams = 5, TotalAssets = 7. In this case we need to load 7 rows (5 rows has both params and holdings and 2 holdings), but `needToFind` would be 0.
This means we can just set `needToFind = TotalAssets`.